docs/proposals/config_data.md

```diff
@@ -0,0 +1,319 @@
+
+## Abstract
+
+This proposal proposes a generic configuration object named ``ConfigData`` stored inside ``etcd`` that
+stores data used for the configuration of ``Kubernetes`` components.
+
+The main focus points of this proposal are:
+
+* Solving dynamic configuration problem of components.
+* Encapsulate configuration information to decrease component configuration complexity.
+* Create a flexible configuration model for ``Kubernetes`` components.
+* Increase overall system quality.
+
+## Motivation
+
+Currently command-line flags and environment variables are used for component configurations in ``Kubernetes``
+which is complicated and error-prone. This approach is not suitable for a distributed computing environment.
+Specifically the problems with the current approach are:
+
+* Complicated deployment orchestration
+* Hard to synchronize across replicated components.
+* Hard to change dynamically as needed by continuous services.
+* Hard to change in different versions that may support different configuration options.
+
+## Solution
+
+A new ``ConfigData`` resource is proposed to solve the aforementioned problems
+that is stored in ``etcd`` and distributed to interested/necessary components
+through the API server. 
+
+The ``ConfigData`` architecture will be very similar to that of ``Secrets``.
+
+``ConfigData`` can either be mounted as a volume or can be injected as environment
+variables to a container through the use of the downward API.
+
+### Dynamic Configuration
+
+Any long-running system has mutating specification over time. In order to facilitate this functionality,
+``ConfigData`` will be versioned and updates will automatically made available to the container (downward API).
+
+``resourceVersion`` (Found in ``ObjectMetadata``) of the ``ConfigData`` object will be updated by the server every time the object is modified.
+After an update, modifications will be made visible to the consumer container. If the consumer uses the ``Data`` pairs
+only for initialization or during starting process, A rolling-update might be necessary to update the components.
+
+It is then the consumer component's responsilibity to make use of the updated data
+once it is made visible (or perform a rolling-update on consumers of that object).
+
+### Advantages
+
+* Reusable across different components.
+* Easy distribution through the API server.
+* Ability to use ``Kubernetes`` authentication and security policies with configuration objects.
+* By leveraging the power of ``/watch`` API along with the new resource, we can change component configurations dynamically.
+* Layer of abstraction that gets rid of the global state currently used for command-line flags and environment variables.
+
+### API Resource
+
+A new resource for ``ConfigData`` will be added to the API:
+
+```go
+	type ConfigData struct {
+		TypeMeta   `json:",inline"`
+		ObjectMeta `json:"metadata,omitempty"`
+
+		Data map[string]string `json:"data,omitempty"`
```

Comment:
What percentage of config data will have json data?  Or YAML?  JSON config
seems low percentage use case.  In OpenShift we started with JSON and
quickly moved to YAML config because it's unreadable/unpredictable to
admins.